src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.InFlightSolution.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal sealed partial class RemoteWorkspace
+    {
+        /// <summary>
+        /// Wrapper around asynchronously produced solution.  The computation for producing the solution will be
+        /// canceled when the number of in-flight operations using it goes down to 0.
+        /// </summary>
+        public sealed class InFlightSolution
+        {
+            private readonly RemoteWorkspace _workspace;
+
+            public readonly Checksum SolutionChecksum;
+
+            private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+            /// <summary>
+            /// Background work to just compute the disconnected solution associated with this <see cref="SolutionChecksum"/>
+            /// </summary>
+            private readonly Task<Solution> _disconnectedSolutionTask;
+
+            /// <summary>
+            /// Optional work to try to elevate the solution computed by <see cref="_disconnectedSolutionTask"/> to be
+            /// the primary solution of this <see cref="RemoteWorkspace"/>.
+            /// </summary>
+            private Task<Solution>? _primaryBranchTask;
```

Comment:
this is an interesting concept in that the InFlightSolution really represents the asynchronous computation of up to two solutions.  The first is just the actual pure 'disconnected' (e.g. not attached to the RemoteWorkspace yet) Solution instance which is always computed.  The second, is the potential computation which takes that solution and tries to make it the primary solution of the worskpace if so requested.  

We always have the first computation. But we may or may not have the second computation depending on the order of operations, and overlapping, of requests from the host to OOP.

Reply:
note: when we get away from Solutions+Workspaces even having a connection in oop, this will go away.